Migrating an endpoint that uses a database. (#1046)
Effect-TS/website: content/src/content/docs/learn/tutorials/hooking-up-a-database.mdx

database. For illustrative purposes, we're going to use SQLite using Bun's
built-in connector, but the technique we'll cover will work for
whatever database you're using, and however you're connecting to it.

First, let's take a look at our app in full with our new endpoint added to
it:

```ts twoslash {8-9,13-17} title=index.ts
import express from "express"

import { HttpRouter, HttpServer, HttpServerResponse } from "@effect/platform"
import { NodeHttpServer, NodeRuntime } from "@effect/platform-node"
import { Effect, Layer } from "effect"
import { createServer } from "node:http"

import { Database } from "bun:sqlite";
const db = new Database("sqlite.db");

const app = express();

app.get("/users", (req, res) => {
  res.send({
    users: db.prepare("SELECT * FROM users").all(),
  });
});

export const health = Effect.sync(() => {
  return HttpServerResponse.text("ok")
})

const router = HttpRouter.empty.pipe(
  HttpRouter.get("/health", health),
)

NodeRuntime.runMain(
  Layer.launch(
    Layer.provide(
      HttpServer.serve(router),
      NodeHttpServer.layer(() => createServer(app), { port: 3000 }),
    ),
  ),

> )

Reviewer: Since you've already taught folks about `.pipe`, you could re-write this to make it flow sequentially:

```ts
HttpServer.server(router).pipe(
  Layer.provide(NodeHttpServer.layer(() => createServer(app), { port: 3000 })),
  Layer.launch,
  NodeRuntime.runMain
)
```

```

To create our example `sqlite.db` file, you can run the following commands:

```shell
$ sqlite3 sqlite.db
sqlite> CREATE TABLE users (id INTEGER PRIMARY KEY AUTOINCREMENT, name TEXT);
sqlite> INSERT INTO users (name) VALUES ("Alice");
sqlite> INSERT INTO users (name) VALUES ("Bob");
```

Then you can run the server with:

```shell
$ bun index.ts
```

> And in other shell we can test the endpoint:

Reviewer: ```suggestion
And in another shell we can test the endpoint:
```

})
```

And this causes the type mismatch error we saw earlier. Let's fix it.

## More pipes

We saw `pipe` used in our `HttpRouter`, and we're going to use it again here.
The `pipe` function is fundamental to Effect, you're going to see it a lot.

Let's express our `users` handler in a different way:

```ts twoslash
import { HttpServerResponse, HttpRouter } from "@effect/platform"
import { Effect, Layer } from "effect"
import { createServer } from "node:http"
import { Database } from "bun:sqlite";
const db = new Database("sqlite.db");
export const health = Effect.sync(() => {
  return HttpServerResponse.text("ok")
})
//---cut---
const users = Effect.sync(() => {
  return { users: db.prepare("SELECT * FROM users").all() }
}).pipe(
  (body) => HttpServerResponse.json(body)
)

const router = HttpRouter.empty.pipe(
  HttpRouter.get("/health", health),
  HttpRouter.get("/users", users),
)
```

This no longer gives us a type error. We've split our computation into two
parts: fetching the data with `Effect.sync`, and serializing it with
`HttpServerResponse.json`. The call to `pipe` takes the result of the first
Effect and passes it to `HttpServerResponse.json`, flattening any nesting in

> the process.

Reviewer: This is not correct - `body` in this case has the type `Effect<{ users: Array<string> }>`, which we're now passing to `HttpServerResponse.json`. Because `HttpServerResponse.json` takes data of type `unknown`, you're going to attempt to serialize an `Effect` instead of the `body`. 

The reason you don't get a type error here is because the input to `HttpServerResponse.json` is `unknown`, which will allow any value.

At this point, you may choose to introduce the concept of sequencing `Effect`s.

```ts
const users = Effect.sync(() => {
  return { users: db.prepare("SELECT * FROM users").all() }
}).pipe(
  // Similar to `Effect.map`, `Effect.flatMap` receives the value of the previous effect 
  // but instead of returning a new _value_, you return a new _Effect_
  Effect.flatMap((body) => HttpServerResponse.json(body))
  // Or alternatively `Effect.andThen((body) => ...)` if you prefer that naming
)
// `users` is type `Effect<HttpServerResponse, HttpBodyError>`
```

or you can introduce `Effect.suspend`, which is the same as `Effect.sync`, except instead of taking a side effect that returns a value, it takes a side effect that returns an `Effect`:

```ts
const users = Effect.suspend(() => {
  // Side effect
  const body = { users: db.prepare("SELECT * FROM users").all() }
  // return an effect
  return HttpServerResponse.json(body)
})
// `users` is type `Effect<HttpServerResponse, HttpBodyError>`
```

Author: With the code as-is, e.g.

```typescript
const users = Effect.sync(() => {
  return {
    users: db.prepare("SELECT * FROM users").all(),
  };
}).pipe((body) => HttpServerResponse.json(body))
```

If I make a request to `localhost:3000/users` I get the following:

```
❯ http get "localhost:3000/users"
HTTP/1.1 200 OK
Content-Length: 57
Content-Type: application/json; charset=utf-8
Date: Fri, 07 Feb 2025 10:15:32 GMT
ETag: W/"39-6Qu85qIU12mhgGYWwHErQVfJRxI"
X-Powered-By: Express

{
    "users": [
        {
            "id": 1,
            "name": "Alice"
        },
        {
            "id": 2,
            "name": "Bob"
        }
    ]
}
```

So I made the assumption it was doing something clever with the `Effect` it receives. 

The result of using flatMap, e.g.

```typescript
const users = Effect.sync(() => {
  return {
    users: db.prepare("SELECT * FROM users").all(),
  };
}).pipe(Effect.flatMap(HttpServerResponse.json));
```

Is identical.

```
❯ http get "localhost:3000/users"
HTTP/1.1 200 OK
Content-Length: 57
Content-Type: application/json; charset=utf-8
Date: Fri, 07 Feb 2025 10:18:23 GMT
ETag: W/"39-6Qu85qIU12mhgGYWwHErQVfJRxI"
X-Powered-By: Express

{
    "users": [
        {
            "id": 1,
            "name": "Alice"
        },
        {
            "id": 2,
            "name": "Bob"
        }
    ]
}
```

Reviewer 2: Looks like a bug in the impl of `json` probably something uses andThen and it ends up flattening when it shouldn't, cc @tim-smart

Author: Sorry, disregard what I'm saying here. I feel silly. I had forgotten to delete the Express handler for the route, and that's what was running. Not the Effect code. There's no bug to fix here, it was failing as expected.

  HttpRouter.get("/health", health),
  HttpRouter.get("/users", users),
)
```

This doesn't work, as we can see from the error. The reason is that
`HttpServerResponse.json` is an operation that might fail. Not all JavaScript
objects can be converted to JSON.

```ts twoslash
JSON.stringify({ big: 10n })
//=> TypeError: JSON.stringify cannot serialize BigInt.
```

For this reason, `HttpServerResponse.json` returns an `Effect` that can fail.
We didn't see this earlier because `HttpServerResponse.text` can't fail, and
so just returns an `HttpServerResponse` directly.

## Error handling in Effect

When we introduced the `Effect` type in the previous section, we dealt
exclusively with Effects that couldn't fail. That won't get us very far if
we're building anything non-trivial.

`Effect` is a generic type that takes 3 arguments:

```typescript
Effect<Success, Error, Requirements>
```

`Success` is the type the Effect returns when it succeeds, `Error` is the type
it returns when it fails, and we'll talk about `Requirements` a bit later.

When we created our example Effects in the previous section, these all took on
the type `Effect<string, never, never>`.

```ts twoslash

> import { Effect } from "effect"
> //---cut---
> Effect.sync(() => "Hello, world!")
> //=> Effect.Effect<string, never, never>

Reviewer: Just as an FYI, if you want to use Twoslash to display the type of something you can use the `// ^?` query operator:

```suggestion
import { Effect } from "effect"
//---cut---
const _ = Effect.sync(() => "Hello, world!")
//        ^?
```

I see you're using it in the snippet below, so I assume this was deliberate 😄

```

This means they will return a `string` on success, and never fail.

The Effect that `HttpServerResponse.json` returns is of type
`Effect<HttpServerResponse, HttpBodyError, never>`. So when we try to return
it from our `users` handler, we end up wrapping one Effect in another.

```ts twoslash
import { HttpServerResponse } from "@effect/platform"
import { Effect, Layer } from "effect"
import { createServer } from "node:http"
import { Client } from "pg";
const db = new Client({ user: "postgres" });
//---cut---
const users = Effect.promise(async () => {
//    ^?
  const { rows } = await db.query("SELECT * FROM users");
  return HttpServerResponse.json({ users: rows });
});
```

This is what the type error from earlier was trying to tell us. Here's the
error again:

```text wrap
Argument of type 'Effect<Effect<HttpServerResponse, HttpBodyError, never>, never, never>' is not assignable to parameter of type 'Handler<never, never>'
```

> Under the hood, a `Handler<never, never>` is an `Effect<HttpServerResponse,
> never, never>`. Because `Effect<HttpServerResponse, HttpBodyError, never>` is
> not an `HttpServerResponse`, the type checker gets upset with us.

Reviewer: ```suggestion
Under the hood, a `Handler<never, never>` is just a type alias for `Effect<HttpServerResponse,
never, never>`. Because `Effect<HttpServerResponse, HttpBodyError, never>` is
not an `HttpServerResponse`, the type checker gets upset with us.
```

## Fixing our nested Effects with `flatMap`

It's possible to fix this with a call to `Effect.flatMap`:

```ts twoslash
import { HttpServerResponse } from "@effect/platform"
import { Effect, Layer } from "effect"
import { createServer } from "node:http"
import { Client } from "pg";
const db = new Client({ user: "postgres" });
//---cut---
const users = Effect.flatMap(
  Effect.promise(async () => {
    const { rows } = await db.query("SELECT * FROM users");
    return { users: rows };
  }),
  HttpServerResponse.json
);
```

But it is more idiomatic in Effect to use `pipe`:

```ts twoslash
import { HttpServerResponse } from "@effect/platform"
import { Effect, Layer } from "effect"
import { createServer } from "node:http"
import { Client } from "pg";
const db = new Client({ user: "postgres" });
//---cut---
const users = Effect.promise(
  () => db.query("SELECT * FROM users")
).pipe(
  Effect.map(({ rows }) => ({ users: rows })),
  Effect.flatMap(HttpServerResponse.json)
);
```

This no longer gives us a type error. We've split our computation into parts:
fetching the data with `Effect.promise`, manipulating its shape with

> `Effect.map`, and serializing it with `HttpServerResponse.json`. The
> `Effect.flatMap` at the end care of the accidental `Effect` nesting.
> 

Reviewer: ```suggestion
`Effect.map`, and serializing it with `HttpServerResponse.json`. The
`Effect.flatMap` at the end takes care of the accidental `Effect` nesting we had in our previous implementation.

```

const mul = (x: number) => (y: number) => y * x
const sub = (x: number) => (y: number) => y - x

const result = pipe(
  1,
  add(2),
  mul(3),
  sub(1)
)
console.log(result)
//=> 8
```

`add`, `mul`, and `sub` are all functions that return functions. So `add(2)`
returns a function that can add 2 to a given number: `add(2)(1)` returns 3.
Our pipeline starts with 1, then passes it through `add(2)`, `mul(3)`, and
`sub(1)`, resulting in 8.

One of the powers this gives you is the ability to add things easily into
the pipeline.

```ts twoslash {6-9,14,16,18}
import { pipe } from "effect"

const add = (x: number) => (y: number) => y + x
const mul = (x: number) => (y: number) => y * x
const sub = (x: number) => (y: number) => y - x
const log = (s: string) => (y: number) => {
  console.log(s, y)
  return y
}

const result = pipe(
  1,
  add(2),
  log("1 + 2 ="),
  mul(3),
  log("  * 3 ="),
  sub(1),
  log("  - 1 ="),
)
```

This prints the following, with `result` still being 8:

```
1 + 2 = 3
  * 3 = 9
  - 1 = 8
```

As we migrate more complex endpoints, we'll see how this ability to "tap into"
the pipeline is really useful.

## Adding error handling to our database query

> All this talk of error handling might have gotten you thinking: can't our
> database query fail? It can! If the query is malformed, `db.query` will throw an
> error. When that happens, at the moment our endpoint will return a 500 error to
> the client. Change the SQL query to `SELECT * FROM nonexistnet` and try it:

Reviewer: ```suggestion
All this talk of error handling might have gotten you thinking: can't our
database query fail? It can! If the query is malformed, `db.query` will throw an
error. When that happens, our endpoint will currently return a 500 error to
the client. Change the SQL query to `SELECT * FROM nonexistent` and try it:
```

you're passing some data through a sequence of functions.

Take this example:

```ts twoslash
import { pipe } from "effect"

const add = (x: number) => (y: number) => y + x
const mul = (x: number) => (y: number) => y * x
const sub = (x: number) => (y: number) => y - x

const result = pipe(
  1,
  add(2),
  mul(3),
  sub(1)
)
console.log(result)
//=> 8
```

`add`, `mul`, and `sub` are all functions that return functions. So `add(2)`
returns a function that can add 2 to a given number: `add(2)(1)` returns 3.
Our pipeline starts with 1, then passes it through `add(2)`, `mul(3)`, and
`sub(1)`, resulting in 8.

One of the powers this gives you is the ability to add things easily into
the pipeline.

```ts twoslash
import { pipe } from "effect"

const add = (x: number) => (y: number) => y + x
const mul = (x: number) => (y: number) => y * x
const sub = (x: number) => (y: number) => y - x
const log = (s: string) => (y: number) => {
  console.log(s, y)
  return y
}

const result = pipe(
  1,
  add(2),
  log("1 + 2 ="),
  mul(3),
  log("  * 3 ="),
  sub(1),
  log("  - 1 ="),
)
```

This prints the following, with `result` still being 8:

```
1 + 2 = 3
  * 3 = 9
  - 1 = 8
```

> As we migrate more complex endpoints, we'll see how this ability to "tap into"

Author: Thank you 🙏 

I'm trying to use language that might be familiar to people coming from other languages, e.g. Ruby has a `tap` method that works similarly: https://ruby-doc.org/3.4.1/Kernel.html#method-i-tap.

---
type: Tutorial
title: Hooking up a database
tags:
  - some
  - tags
sidebar:
  order: 3
---

import { Aside } from "@astrojs/starlight/components"

The next endpoint we're going to write is one that lists all of the users in a
database. For this tutorial I'm going to use Postgres running in Docker as our
database. If you don't have Docker installed, I recommend following the
[official getting started guide](https://www.docker.com/get-started/).

When you're ready, you can start a Postgres container with:

```shell showLineNumbers=false
docker run -d -e POSTGRES_HOST_AUTH_METHOD=trust -p 5432:5432 postgres:17
```

<Aside type="note" title="I can't use Docker!">
If you aren't able to run Docker locally, for example because your work's laptop
policy doesn't allow it, an alternative is to use a hosted Postgres service.
For example, [Neon](https://neondb.com/) offers a free tier that should be
more than enough for this tutorial. They also offer a web-based UI for creating

> tables and rows.

Reviewer: While I understand wanting to point users in the right direction, I'm not sure we should direct them to a _specific_ hosted service.

```

To exit `psql`, type `\q` and press enter.

## The `/users` endpoint

To talk to Postgres we'll need a client library. We're going to use `pg` for
this tutorial. You can install it with:

```shell showLineNumbers=false
bun add pg @types/pg
```

To start with, we're going to hardcode our database connection details. Later
on we'll learn how to use Effect to manage our app's configuration.

Here's the code for our app in full, with a new `/users` endpoint added in
Express:

```ts twoslash ins={6,8,11-14,25} title=index.ts showLineNumbers=false
import express from "express"

import { NodeHttpServer, NodeRuntime } from "@effect/platform-node"
import { HttpRouter, HttpServerResponse } from "@effect/platform"
import { Effect } from "effect"
import { Client } from "pg"

const db = new Client({ user: "postgres" })
const app = express()

app.get("/users", async (req, res) => {
  const { rows } = await db.query("SELECT * FROM users")
  res.send({ users: rows })
})

const health = Effect.sync(() => {
  return HttpServerResponse.text("ok")
})

const router = HttpRouter.empty.pipe(
  HttpRouter.get("/health", health)
)

const main = Effect.gen(function* () {

>   yield* Effect.promise(() => db.connect())

Reviewer: ```suggestion
  // Make sure we connect to the database before starting the server
  yield* Effect.promise(() => db.connect())
```

Might be worth adding a comment just for this  snippet since user's might be a bit confused why this is required? Idk

```ts twoslash
import { HttpServerResponse } from "@effect/platform"
import { Effect, Layer } from "effect"
import { createServer } from "node:http"
import { Client } from "pg"
const db = new Client({ user: "postgres" })
//---cut---
const users = Effect.tryPromise(
  () => db.query("SELECT * FROM users")
).pipe(
  Effect.map(({ rows }) => ({ users: rows })),
  Effect.flatMap(HttpServerResponse.json)
)
```

This changes the return type from:

```typescript
Effect<HttpServerResponse, HttpBodyError, never>
```

To:

```typescript
Effect<HttpServerResponse, UnknownException | HttpBodyError, never>
```

`Effect.promise` is assumed to always succeed. `Effect.tryPromise` is assumed
to maybe throw an exception, but because TypeScript doesn't give us any way
to know _what_ exception might be thrown, Effect defaults to `UnknownException`.

We can improve on this:

```ts twoslash {1,4-5}
import { HttpServerResponse } from "@effect/platform"
import { Effect, Layer } from "effect"
import { createServer } from "node:http"
import { Client} from "pg"
const db = new Client({ user: "postgres" })
//---cut---
class UserFetchError extends Error {}

> 

Reviewer: For a variety of reasons, I would advise against this pattern. 

For one, the lack of a discriminator in the type (e.g. a `_tag`) means that another error which extends `Error` can be directly assignable to the `UserFetchError`: https://effect.website/play#55cd56ec464d.

My recommendation would be to either:

1. Take the time to introduce creating your own domain-specific errors more properly here via `Data.TaggedError` https://effect.website/docs/data-types/data/#taggederror
2. Just make the error an interface in this case for now if you want to keep it simple

```ts
interface UserFetchError {
  readonly _tag: "UserFetchError"
  readonly message: string
  readonly cause: unknown
}
```

I personally think it might be good to go for the first option given that `Data.TaggedError` will automatically attach a `_tag` to our error type and also extends `Error` for us by default, but I leave that decision to you.

Author: I agonised over this, because tags do feel like they deserve special attention and I'll soon get to them when I talk about services. I wanted to wait for a place I _needed_ to use them in error handling to introduce them. I also don't want to wander too far away from the purpose of this section: to migrate an endpoint that calls an external API.

This section in its entirety is still very early, probably the earliest I've pushed a section up for review. I'm not tightly committed to any part of it. If you think we really should introduce tags here, I don't think it would be impossible to do so. I do think we'll have to cut something else to make room for them, though, and it might have to be services.

Reviewer: Totally understand, and honestly I don't want to impose my opinion too much on the direction you take here.

My only advice would be that if you want to avoid introducing `Data.TaggedError`, etc., maybe just add a `_tag` property or something to your custom error type and do another one of those "this is important but we'll discuss it later" things. Just to ensure you have a discriminator for the type in place, and you don't create an unintentional footgun with multiple error types that are assignable to one another.

const users = Effect.tryPromise({
  try: () => db.query("SELECT * FROM users"),
  catch: (cause) => new UserFetchError("failed to fetch users", { cause }),
}).pipe(
  Effect.map(({ rows }) => ({ users: rows })),
  Effect.flatMap(HttpServerResponse.json)
)
```

We're using a different variant of `Effect.tryPromise` to give us more control
over the type of the error that is returned. Becuase it's difficult to know what
errors may be thrown by third-party libraries, it's idiomatic to create new
error types that wrap the original error.

Now our effect has the type:

```typescript
Effect<HttpServerResponse, UserFetchError | HttpBodyError, never>
```

Now our program is aware of what can go wrong _at the type level_. This
information is completely missing if you're using exceptions in TypeScript.
With this information we can ensure, and get TypeScript to verify for us,
that things do not go wrong in unexpected ways.

```ts twoslash
// @errors: 2375
import { Effect } from "effect"

const effect = Effect.succeed(1).pipe(
  Effect.tryMap({
    try: (n) => n + 1,
    catch: () => new Error(),
  }),
)

> const noError: Effect.Effect<number, never, never> = effect

Reviewer: This transition felt a bit abrupt to me. To me, it might feel more natural if you insert one of the "aside"-esque comments you've had previously (recall the `pipe` aside). Perhaps "An aside on errors in Effect" or something?

Also, before now `Effect.tryMap` has not even been mentioned but now we're using it all of a sudden. 

Might be worth using the user fetch code from above for these samples instead of the random `tryMap` code, given that you also insert it into the final snippet later anyways.

Author: I did wonder about whether to explicitly call out `Effect.tryMap` and landed on not doing so. It felt like because we've already covered `Effect.map`, and we looked at `Effect.try` in the previous section, the reader wouldn't flinch at `Effect.tryMap` being casually used without explanation.

A small nod to it might be the way to go, though. I need to start being careful, I've not spent more time with Effect than the reader and am likely prone to thinking they'll understand things they might not. 😅

```

The above code fails because we try to assign an Effect that can fail to a
variable that expects an Effect that can't fail. We can fix this error in a
few ways.

By providing a default value:

```ts twoslash {8}
import { Effect } from "effect"

const effect = Effect.succeed(1).pipe(
  Effect.tryMap({
    try: (n) => n + 1,
    catch: () => new Error(),
  }),
  Effect.orElseSucceed(() => 0),
)

const noError: Effect.Effect<number, Error, never> = effect
```

Or by converting the error into what Effect calls a "defect":

```ts twoslash {8}
import { Effect } from "effect"

const effect = Effect.succeed(1).pipe(
  Effect.tryMap({
    try: (n) => n + 1,
    catch: () => new Error(),
  }),
  Effect.orDie
)

const noError: Effect.Effect<number, Error, never> = effect
```

This effectively recreates the behaviour of throwing an exception. Defects are
critical errors that you don't want to handle. If you were to create a defect in

> an Effect handler, the result would be a 500 error returned to the user.

Reviewer: Might be worth one more sentence or something here providing a clear distinction between "expected" errors, which are part of the program's domain, and "defects" which are critical errors (as you explained).
